pygeo/mphys/mphys_dvgeo.py

```diff
@@ -997,34 +1040,45 @@ def compute_jacvec_product(self, inputs, d_inputs, d_outputs, mode):
-                            # xdot = DVGeo.totalSensitivityTransProd(dout, ptSetName)
-                            xdot = DVGeo.totalSensitivity(dout, ptSetName)
-
-                            # loop over dvs and accumulate
-                            xdotg = {}
-                            for k in xdot:
-                                # check if this dv is present
-                                if k in d_inputs:
-                                    # do the allreduce
-                                    # TODO reove the allreduce when this is fixed in openmdao
-                                    # reduce the result ourselves for now. ideally, openmdao will do the reduction itself when this is fixed. this is because the bcast is also done by openmdao (pyoptsparse, but regardless, it is not done here, so reduce should also not be done here)
-                                    xdotg[k] = self.comm.allreduce(xdot[k], op=MPI.SUM)
-
-                                    # accumulate in the dict
-                                    # TODO
-                                    # because we only do one point set at a time, we always want the 0th
-                                    # entry of this array since dvgeo always behaves like we are passing
-                                    # in multiple objective seeds with totalSensitivity. we can remove the [0]
-                                    # once we move back to totalSensitivityTransProd
-                                    d_inputs[k] += xdotg[k][0]
+                            if doFwd:
+                                d_outputs[ptSetName] += DVGeo.totalSensitivityProd(seeds, ptSetName)
+                            elif doRev:
+                                # TODO totalSensitivityTransProd is broken. does not work with zero surface nodes on a proc
+                                # xdot = DVGeo.totalSensitivityTransProd(dout, ptSetName)
+                                xdot = DVGeo.totalSensitivity(seeds, ptSetName)
+
+                                # loop over dvs and accumulate
+                                xdotg = {}
+                                for k in xdot:
+                                    # check if this dv is present
+                                    if k in d_inputs:
+                                        # do the allreduce
+                                        # TODO remove the allreduce when this is fixed in openmdao
+                                        # reduce the result ourselves for now. ideally, openmdao will do the reduction itself when this is fixed. this is because the bcast is also done by openmdao (pyoptsparse, but regardless, it is not done here, so reduce should also not be done here)
+                                        xdotg[k] = self.comm.allreduce(xdot[k], op=MPI.SUM)
+
+                                        # accumulate in the dict
+                                        # TODO
```

Comment:
If this is working well I think we can clean up this comment as well. If there's actually plans to fix and move back to `totalSensitivityTransProd` then we should create and issue and reference this code in the issue. Otherwise this comment will probably get lost to the ether.